src/common/utils.h

```diff
@@ -63,6 +63,36 @@ inline bool ContainsOnlyStorage(const StorageTypeVector& vstorage,
   return false;
 }
 
+/*! \brief returns true if all storage types in `vstorage` are the same as target `stype1`
+ *         or `stype2'. Sets boolean if both found.
+ *         false is returned for empty inputs.
+ */
+inline bool ContainsOnlyStorage(const StorageTypeVector& vstorage,
```

Comment:
We are having too many of the these utils. Its starting to look like something is wrong with the design.

Reply:
The new design may check a lot of permutations of inputs/outputs in order to make the decision to fall back or not.  This is the foundation of the new infer storage type design which is already in master.  Not sure what the expectation is in that case.

Reply:
We should revisit this after the 1.0 release.

Also need to extract sparse related code into separate file/folder. Currently it's all mingled together and too complicated for people who just want to understand elementwise ops